### ISSUE TITLE
How can I disable flash_attn during training?Should I add attn_implementation='eager' in the llm section of the config?

### ISSUE THREAD

**Issue by 330205812** [InternLM/xtuner#732]
model = dict(
    llm=dict(
        attn_implementation='eager',
        pretrained_model_name_or_path=
        '/home/zhangwq/project/QA_SERVICE/LLM_model/internlm2/internlm2-7b',
        torch_dtype='torch.float16',
        trust_remote_code=True,
        type='transformers.AutoModelForCausalLM.from_pretrained')）

When I add attn_implementation='eager' in the llm dictionary, the training speed is almost the same with before.

**Comment by HIT-cwh** [comment]
Hi @330205812 ! Could you provide the following details:
1. xtuner version
2. `max_length` in config
3. `pack_to_max_length` in config

**Comment by 330205812** [comment]
> Hi @330205812 ! Could you provide the following details:
> 
> 1. xtuner version
> 2. `max_length` in config
> 3. `pack_to_max_length` in config

xtuner version is 0.1.18 and 0.1.19
max_length = 512
pack_to_max_length = False

**Comment by HIT-cwh** [comment]
If flash_attn is installed, XTuner version 0.1.18 utilizes flash_attn for training, irrespective of whether attn_implementation is specified in the configuration. This problem has been addressed in version 0.1.19 (refer to [pr#609](https://github.com/InternLM/xtuner/pull/609)).

With XTuner 0.1.19, if you prefer not to utilize flash_attn for training, you can specify `attn_implementation='eager' # or 'fdpa'` in the configuration.

BTW. The acceleration provided by flash_attn becomes more pronounced when dealing with longer training data such as 32k.

**Comment by 330205812** [comment]
> If flash_attn is installed, XTuner version 0.1.18 utilizes flash_attn for training, irrespective of whether attn_implementation is specified in the configuration. This problem has been addressed in version 0.1.19 (refer to [pr#609](https://github.com/InternLM/xtuner/pull/609)).
> 
> With XTuner 0.1.19, if you prefer not to utilize flash_attn for training, you can specify `attn_implementation='eager' # or 'sdpa'` in the configuration.
> 
> BTW. The acceleration provided by flash_attn becomes more pronounced when dealing with longer training data such as 32k.

Thanks for your kind reply. The following is the log of internlm2 I trained with xtuner v0.1.19. The previous config did not add attn_implementation='eager'  in config (i.e., use flash_attn by default), and the later did not add attn_implementation='eager' (i.e., do not use)

(Added attn_implementation='eager')
/05/30 01:36:41 -mmenginee-info-iter (train) [10/1839] lr: 2.6668e-04 eta: 0:52:26 time: 1.7201 data_time: 0.0139 memory: 22978 loss: 1.2352
2024/05/30 01:36:57 - mmengine-INFO-Iter (train) [20/1839] lr: 5.6297e-04 eta: 0:49:55 time: 1.5731 data_time: 0.0163 memory: 23578 loss: 1.0675
2024

(Not added)
2024/05/29 13:16:20 - mmenginee-info-iter (train) [10/1839] lr: 2.6668e-04 eta: 0:49:19 time: 1.6181 data_time: 0.0170 memory: 22974 loss: 1.2001
2024/05/29 13:16:34 - mmenginee-info-iter (train) [20/1839] lr: 5.6297e-04 eta: 0:46:15 time: 1.4340 data_time: 0.0154 memory: 23574 loss: 1.1257

Why are the two training speeds the same?

**Comment by 330205812** [comment]
> If flash_attn is installed, XTuner version 0.1.18 utilizes flash_attn for training, irrespective of whether attn_implementation is specified in the configuration. This problem has been addressed in version 0.1.19 (refer to [pr#609](https://github.com/InternLM/xtuner/pull/609)).
> 
> With XTuner 0.1.19, if you prefer not to utilize flash_attn for training, you can specify `attn_implementation='eager' # or 'sdpa'` in the configuration.
> 
> BTW. The acceleration provided by flash_attn becomes more pronounced when dealing with longer training data such as 32k.

Thanks for your kind reply. The following is the log of internlm2 I trained with xtuner v0.1.19. The previous config added attn_implementation='eager' in config (i.e., use flash_attn by default), and the later did not added attn_implementation='eager' (i.e., do not use flash_attn)

(Added attn_implementation='eager')
/05/30 01:36:41 -mmenginee-info-iter (train) [10/1839] lr: 2.6668e-04 eta: 0:52:26 time: 1.7201 data_time: 0.0139 memory: 22978 loss: 1.2352
2024/05/30 01:36:57 - mmengine-INFO-Iter (train) [20/1839] lr: 5.6297e-04 eta: 0:49:55 time: 1.5731 data_time: 0.0163 memory: 23578 loss: 1.0675
2024

(Not added)
2024/05/29 13:16:20 - mmenginee-info-iter (train) [10/1839] lr: 2.6668e-04 eta: 0:49:19 time: 1.6181 data_time: 0.0170 memory: 22974 loss: 1.2001
2024/05/29 13:16:34 - mmenginee-info-iter (train) [20/1839] lr: 5.6297e-04 eta: 0:46:15 time: 1.4340 data_time: 0.0154 memory: 23574 loss: 1.1257

Why are the two training speeds the same?

**Comment by HIT-cwh** [comment]
Could you provide the log file of these two experiments?

**Comment by 330205812** [comment]
> Could you provide the log file of these two experiments?

Thank you very much for your kindness reply, attach log files for both experiments below
[20240529_131400.log](https://github.com/InternLM/xtuner/files/15496012/20240529_131400.log)
[20240530_013428(attn_implementation='eager').log](https://github.com/InternLM/xtuner/files/15496013/20240530_013428.attn_implementation.eager.log)

**Comment by 330205812** [comment]
> It appears that the iteration time of the model with flash attention is approximately 1.38 seconds, whereas that of the model with eager attention is roughly 1.55 seconds. Flash attention holds a slight advantage. The acceleration provided by flash_attn becomes more pronounced when dealing with longer training data such as 32k, achieved by configuring max_length to 32768, enabling pack_to_max_length, and setting batch_size to 1.

Thank you again for your heartfelt advice. Anyway, eager attention is definitely not using flash attn2, right? Because I was comparing the training performance of different chips, I used xtuner v0.1.18 to train internlm2 on dedicated chips without flash attn2 installed, and trained at about 40s. If it is determined that eager attention is not using flash attn2 (even if it is installed), then I can conclude that there is something wrong with our underlying operator.

The attached log file for the experiment using xtuner v0.1.18 on dedicated chips is below.
[20240527_173028.log](https://github.com/InternLM/xtuner/files/15496788/20240527_173028.log)

**Comment by HIT-cwh** [comment]
There is another way to intuitively determine whether flash attn2 is used for xtuner v0.1.19. If `dispatch internlm2 attn forward` is printed in the log, then flash attn2 is used.
![image](https://github.com/InternLM/xtuner/assets/41630003/f7d1ea08-3f0e-438c-a2f4-9803e97faf23)

**Comment by 330205812** [comment]
> There is another way to intuitively determine whether flash attn2 is used for xtuner v0.1.19. If `dispatch internlm2 attn forward` is printed in the log, then flash attn2 is used. ![image](https://private-user-images.githubusercontent.com/41630003/335127583-f7d1ea08-3f0e-438c-a2f4-9803e97faf23.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTcwNjIzMzgsIm5iZiI6MTcxNzA2MjAzOCwicGF0aCI6Ii80MTYzMDAwMy8zMzUxMjc1ODMtZjdkMWVhMDgtM2YwZS00MzhjLWEyZjQtOTgwM2U5N2ZhZjIzLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTMwVDA5NDAzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAxYWE4N2Q5MDRiY2JmMDgwN2QyYWQ1ODFjYmJhNTVkNDY2ZjZkOWUzNWM3NDk5NDFlZTZlMWI5YzU5N2ZhYWQmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.GMR3cg90iyxVXUH_TMUpydxm3-8Ps1KSRKeQTtW8Gqs)

Thank you for your valuable advice. It appears that some issues with one of the operators on our dedicated chips is significantly slowing down the training process.
